src/frontend/src/helpers/Login.ts

```diff
@@ -0,0 +1,59 @@
+import * as $ from "jquery";
+export {Login, LoginError, Connection}
+
+class LoginError extends Error {
+  constructor(msg: string,
+              public username?: string,
+              public status?: number,
+              public statusText?: number) {
+    super(msg);
+  }
+}
+
+class Login {
+  public uri: string;
+
+  constructor(uri?: string) {
+    this.uri = uri || "https://www.student.cs.uwaterloo.ca/~cs136/seashell/cgi-bin/login2.cgi";
+  };
+
+  public async login(user: string, password: string, rebootBackend?: boolean): Promise<Connection> {
+    try {
+      const response = await $.ajax({
+        url: this.uri,
+        type: "POST",
+        data: {
+          "u": user,
+          "p": password,
+          "reset": !! rebootBackend
+        },
+        dataType: "json"
+      });
+      return new Connection(user,
+                            response.key,
+                            response.host,
+                            response.port,
+                            response.pingPort);
+    } catch (ajax) {
+      const code       = ajax.responseJSON.error.code;
+      const msg        = ajax.responseJSON.error.message;
+      const status     = ajax.status;
+      const statusText = ajax.statusText;
+      throw new LoginError(`Login failure (${code}): ${msg}`, user, status, statusText);
+    }
+  }
+}
+
+
+class Connection {
```

Comment:
Yeah this is the class that can be stored/retrieved in cookies.
I think the constructor arguments is a shorthand for 
```
public username;
public key;
...
this.username = username;
this.key = key;
...
```